File: rover.go
package main

import (
	"github.com/MarsRover/facing"
	moveEnum "github.com/MarsRover/move"
	turnEnum "github.com/MarsRover/turn"
	"math"
)

type marsRover struct {
	position position
	facing   facing.Direction
	world    world
}

func NewMarsRover(initialPosition position, facing facing.Direction, world world) *marsRover {
	return &marsRover{
		position: initialPosition,
		facing:   facing,
		world:    world,
	}
}

func (rover *marsRover) Move(movements []moveEnum.Move) {
	for _, move := range movements {
		switch move {
		case moveEnum.Forward:
			if rover.facing == facing.North {
				rover.position.y++

				if int(math.Abs(float64(rover.position.y))) > rover.world.length {
					rover.position.y--
					rover.position.y *= -1
				}
			} else if rover.facing == facing.South {
				rover.position.y--

				if int(math.Abs(float64(rover.position.y))) > rover.world.length {
					rover.position.y++
					rover.position.y *= -1
				}
			} else if rover.facing == facing.East {
				rover.position.x++

				if int(math.Abs(float64(rover.position.x))) > rover.world.width {
					rover.position.x--
					rover.position.x *= -1
				}
			} else if rover.facing == facing.West {
				rover.position.x--

				if int(math.Abs(float64(rover.position.x))) > rover.world.width {
					rover.position.x++
					rover.position.x *= -1
				}
			}
		case moveEnum.Backward:
			if rover.facing == facing.North {
				rover.position.y--
			} else if rover.facing == facing.South {
				rover.position.y++
			} else if rover.facing == facing.East {
				rover.position.x--
			} else if rover.facing == facing.West {
				rover.position.x++
			}
		}
	}
}

func (rover *marsRover) Turn(turns []turnEnum.Turn) {
	for _, turn := range turns {
		switch turn {
		case turnEnum.Left:
			if rover.facing == facing.North {
				rover.facing = facing.West
			} else if rover.facing == facing.South {
				rover.facing = facing.East
			} else if rover.facing == facing.East {
				rover.facing = facing.North
			} else if rover.facing == facing.West {
				rover.facing = facing.South
			}
		case turnEnum.Right:
			if rover.facing == facing.North {
				rover.facing = facing.East
			} else if rover.facing == facing.South {
				rover.facing = facing.West
			} else if rover.facing == facing.East {
				rover.facing = facing.South
			} else if rover.facing == facing.West {
				rover.facing = facing.North
			}
		}
	}
}

File: MarsRover/Rover.cs
namespace MarsRover
{
    public class Rover
    {
        public Position Position { get; set; }
        public Direction Direction { get; set; }
        public World World { get; set; }

        public Rover() : this(0, 0)
        {
        }

        public Rover(int xInitPos, int yInitPos) : this(new Position(xInitPos, yInitPos))
        {
        }

        public Rover(Position initPosition, Direction direction = Direction.East) : this(initPosition, direction, new World(0, 1, 0 , 1))
        {
        }

        public Rover(Position initPosition, Direction direction, World world)
        {
            this.Position = initPosition;
            this.Direction = direction;
            this.World = world;
        }

        public void Move(int nrToMove)
        {
            switch (this.Direction)
            {
                case Direction.East:
                    this.Position.X += nrToMove;
                    if (this.Position.X > this.World.XMax) {
                        this.Position.X = (this.Position.X % this.World.XMax) + this.World.XMin;
                    }
                    break;
                case Direction.West:
                    this.Position.X -= nrToMove;
                    if (this.Position.X < this.World.XMin)
                    {
                        this.Position.X = (this.Position.X % this.World.XMax) + this.World.XMax;
                    }
                    break;
                case Direction.North:
                    this.Position.Y += nrToMove;
                    if (this.Position.Y > this.World.YMax)
                    {
                        this.Position.Y = (this.Position.Y % this.World.YMax) + this.World.YMin;
                    }
                    break;
                case Direction.South:
                    this.Position.Y -= nrToMove;
                    if (this.Position.Y < this.World.YMin)
                    {
                        this.Position.Y = (this.Position.Y % this.World.YMax) + this.World.YMax;
                    }
                    break;
            }
        }

        public void TurnLeft()
        {
            switch (this.Direction)
            {
                case Direction.East:
                    this.Direction = Direction.North;
                    break;
                case Direction.North:
                    this.Direction = Direction.West;
                    break;
                case Direction.West:
                    this.Direction = Direction.South;
                    break;
                case Direction.South:
                    this.Direction = Direction.East;
                    break;
            }
        }

        public void TurnRight()
        {
            switch (this.Direction)
            {
                case Direction.East:
                    this.Direction = Direction.South;
                    break;
                case Direction.South:
                    this.Direction = Direction.West;
                    break;
                case Direction.West:
                    this.Direction = Direction.North;
                    break;
                case Direction.North:
                    this.Direction = Direction.East;
                    break;
            }
        }
    }
}
